merge_base.go

```diff
@@ -0,0 +1,211 @@
+	if err == errIsReachable {
+		return []*object.Commit{older}, nil
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	var res []*object.Commit
+	inNewerHistory := isInIndexCommitFilter(newerHistory)
+	resIter := object.NewFilterCommitIter(older, &inNewerHistory, &inNewerHistory)
+	err = resIter.ForEach(func(commit *object.Commit) error {
+		res = append(res, commit)
+		return nil
+	})
+
+	return Independents(res)
+}
+
+// IsAncestor returns true if the candidate commit is ancestor of the target one
+// It returns an error if the history is not transversable
+// It mimics the behavior of `git merge --is-ancestor candidate target`
+func IsAncestor(
+	candidate *object.Commit,
+	target *object.Commit,
+) (bool, error) {
+	_, err := ancestorsIndex(candidate, target)
+	if err == errIsReachable {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// ancestorsIndex returns a map with the ancestors of the starting commit if the
+// excluded one is not one of them. It returns errIsReachable if the excluded commit
+// is ancestor of the starting, or another error if the history is not transversable.
+func ancestorsIndex(excluded, starting *object.Commit) (map[plumbing.Hash]struct{}, error) {
+	if excluded.Hash.String() == starting.Hash.String() {
+		return nil, errIsReachable
+	}
+
+	startingHistory := map[plumbing.Hash]struct{}{}
+	startingIter := object.NewCommitIterBSF(starting, nil, nil)
+	err := startingIter.ForEach(func(commit *object.Commit) error {
+		if commit.Hash == excluded.Hash {
+			return errIsReachable
+		}
+
+		startingHistory[commit.Hash] = struct{}{}
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return startingHistory, nil
+}
+
+// Independents returns a subset of the passed commits, that are not reachable the others
+// It mimics the behavior of `git merge-base --independent commit...`.
+func Independents(commits []*object.Commit) ([]*object.Commit, error) {
+	// use sortedByCommitDateDesc strategy
+	cleaned := sortByCommitDateDesc(commits...)
+	cleaned = removeDuplicated(cleaned)
+	return independents(cleaned, map[plumbing.Hash]bool{}, 0)
+}
+
+func independents(
+	candidates []*object.Commit,
+	excluded map[plumbing.Hash]bool,
+	start int,
+) ([]*object.Commit, error) {
+	if len(candidates) == 1 {
+		return candidates, nil
+	}
+
+	res := candidates
+	for i := start; i < len(candidates); i++ {
+		from := candidates[i]
+		others := remove(res, from)
+		fromHistoryIter := object.NewCommitIterBSF(from, excluded, nil)
+		err := fromHistoryIter.ForEach(func(fromAncestor *object.Commit) error {
+			for _, other := range others {
+				if fromAncestor.Hash == other.Hash {
+					res = remove(res, other)
+					others = remove(others, other)
+				}
+			}
+
+			if len(res) == 1 {
+				return storer.ErrStop
+			}
+
+			excluded[fromAncestor.Hash] = true
+			return nil
+		})
+
+		if err != nil {
+			return nil, err
+		}
+
+		if len(res) < len(candidates) {
+			return independents(res, excluded, indexOf(res, from)+1)
```

Comment:
Let's avoid recursion? Git histories can get to >1M commits.

Reply:
I'm not sure about understanding your suggestion.
This recursive call precisely reduces the iterations and steps needed to find the independent commits, because once one is found as reachable from the others, it removes it, and jumps to the next one.

Reply:
...or I could try with a `continue`... to make it more explicit how the filtering works :dancer:
(I'll try later)